src/uri.jl

```diff
@@ -1,435 +1,144 @@
-is_url_char(c) =  ((@assert UInt32(c) < 0x80); 'A' <= c <= '~' || '$' <= c <= '>' || c == '\f' || c == '\t')
-is_mark(c) = (c == '-') || (c == '_') || (c == '.') || (c == '!') || (c == '~') ||
-             (c == '*') || (c == '\'') || (c == '(') || (c == ')')
-is_userinfo_char(c) = isalnum(c) || is_mark(c) || (c == '%') || (c == ';') ||
-             (c == ':') || (c == '&') || (c == '+') || (c == '$' || c == ',')
-isnum(c) = ('0' <= c <= '9')
-ishex(c) =  (isnum(c) || 'a' <= lowercase(c) <= 'f')
-is_host_char(c) = isalnum(c) || (c == '.') || (c == '-') || (c == '_') || (c == "~")
-
-"""
-`HTTP.URI(str::String)` => `HTTP.URI`
+# URI
+immutable Offset
+    off::UInt16
+    len::UInt16
+end
+Offset() = Offset(0, 0)
+Base.getindex(A::Vector{UInt8}, o::Offset) = String(A[o.off:(o.off + o.len - 1)])
+Base.isempty(o::Offset) = o.off == 0x0000 && o.len == 0x0000
+==(a::Offset, b::Offset) = a.off == b.off && a.len == b.len
 
-A type representing a uri/url used for resource identification on the web.
-"""
 immutable URI
-    scheme::String
-    host::String
-    port::UInt16
-    path::String
-    query::String
-    fragment::String
-    userinfo::String
-    specifies_authority::Bool
-    URI(scheme,host,port,path,query="",fragment="",userinfo="",specifies_authority=false) =
-            new(scheme,host,UInt16(port == 0 ? (scheme == "http" ? 80 : scheme == "https" ? 443 : 0) : port),
-                path,query,fragment,userinfo,specifies_authority)
+    data::Vector{UInt8}
+    offsets::NTuple{7, Offset}
 end
 
 const URL = URI
 
-==(a::URI,b::URI) = (a.scheme   == b.scheme)   &&
-                    (a.host     == b.host)     &&
-                    (a.port     == b.port)     &&
-                    (a.path     == b.path)     &&
-                    (a.query    == b.query)    &&
-                    (a.fragment == b.fragment) &&
-                    (a.userinfo == b.userinfo)
-
-URI(host, path) = URI("http", host, UInt16(80), path, "", "", "", true)
-
-function URI(uri::URI; kwargs...)
-    nms = fieldnames(URI)
-    args = Pair{Symbol}[(nms[i] => getfield(uri, i)) for i = 1:nfields(HTTP.URI)]
-    for (k, v) in kwargs
-        for (i, arg) in enumerate(args)
-            if arg[1] == k
-                args[i] = Pair(k, v)
-            end
-        end
-    end
-    return URI([arg[2] for arg in args]...)
+function URI(;hostname::String="", path::String="",
```

Comment:
Couple of reasons:

  * In general, the webstack has been poorly maintained, I believe partly because things were too separated out and there was quite a bit of code duplication across the packages (or code in packages where it didn't belong); downstream packages were almost exclusively depending on Requests or HttpServer anyway
  * The URIParser code is pretty out-dated and the API isn't very consistent; while cleaning it up, I ended up going down the rabbit-hole of porting the libhttp-parser code over and so decided to restructure entirely anyway
  * The new representation is quite a bit more efficient now in terms of storage and performance

It might make sense to split some functionality back off in the future, but for now, I'd of the belief that the webstack will be better maintained if issues and dev can be centralized as much as possible.